### PR TITLE
Add gan.py

### PR DESCRIPTION
**Pull request by wangkuiyi**
Add this test-driven development test 

Fix https://github.com/wangkuiyi/fluid/issues/24

### CODE REVIEW

File: python/fluid/gan.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+def create_discriminator_parameters():
+    return {"Dw": fluid.tensor(dim=[20, 2]), "Db": fluid.tensor(dim=[2])}
+
+
+def generate(noise, params):
+    return fluid.fc(noise, params["Gw"], params["Gb"])
+
+
+with fluid.defun("train_discriminator", mb, params):
+    fluid.optimize(
```
Comment (reyoung) [inline python/fluid/gan.py:30]: I am not sure whether we should combine `backward` and `optimize` together. However, if we want to combine them together, I suggest we use `fluid.optimizers.adam()` or `fluid.optimizers.sgd`, instead of `fluid.optimize`

File: python/fluid/gan.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        fluid.mse(
```
Comment (reyoung) [inline python/fluid/gan.py:31]: `fluid.mse` --> `fluid.cross_entropy`

File: python/fluid/gan.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            fluid.softmax(
+                fluid.fc(
+                    generate(noise, g_params), d_params["Dw"], d_params["Db"]),
+                fluid.reshape(True, fluid.sizeof(noise)))), g_params)
+
+generator_params = create_generator_parameters()
+discriminator_params = create_discriminator_parameters()
+
+with fluid.loop(iter=1000):
```
Comment (reyoung) [inline python/fluid/gan.py:46]: Should we support `key-value arguments` for `function`.

File: python/fluid/gan.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    # Read real images
+    real_images = fluid.data("images")
+
+    # Generate fake images
+    noise = fluid.random()
+    fake_images = generate(noise, generator_params)
+
+    # Mix real and fake images
+    mb = fluid.enssamble_minibatch(real_images, True, fake_images, False)
```
Comment (reyoung) [inline python/fluid/gan.py:55]: What is `enssamble_minibatch`? What are these boolean parameters?

Maybe just to use `fluid.concat(real_images, fake_images)`

File: python/fluid/gan.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        params)
+
+with fluid.defun("train_generator", noise, g_params, d_params):
+    fluid.optimize(
+        fluid.mse(
+            fluid.softmax(
+                fluid.fc(
+                    generate(noise, g_params), d_params["Dw"], d_params["Db"]),
+                fluid.reshape(True, fluid.sizeof(noise)))), g_params)
+
```
Comment (reyoung) [inline python/fluid/gan.py:42]: I think the functional style API is cool, but a little bit noising.

Maybe for function only have one input and one output, we can use `a.func()` style API?

It likes

```
fluid.fc(generate(noise, g_params), d_params["Dw"], d_params["Db"])).softmax()
```

File: python/fluid/gan.py (hunk 4 above)
Comment (wangkuiyi) [reply python/fluid/gan.py:55]: Good point. I don't have a good name for it yet. Would love to hear your suggestion -- what I want to do is to form a matrix

| real images | True |
|------|-----|
| fake images | False |

I think it is more than `concat` because I also want to add the labels.

File: python/fluid/gan.py (hunk 1 above)
Comment (wangkuiyi) [reply python/fluid/gan.py:30]: I haven't thought of as much detail. But I feel that we should pack

1. backward, 
1. gradient aggregation, and 
1. parameter update 

in a single function call, so that 

1. we can parallelize them, and
1. we can encapsulate distributed computing details (Kubernetes, MPI, SGE, etc) -- deep learners are anyway less interested in setting up distributed training configurations, so let us take full care of it.

File: python/fluid/gan.py (hunk 5 above)
Comment (reyoung) [reply python/fluid/gan.py:42]: However, it is not important. We might implement a prototype first.

File: python/fluid/gan.py (hunk 3 above)
Comment (wangkuiyi) [reply python/fluid/gan.py:46]: Given that is a Python convention, I don't see a reason that we should prohibit from using named parameters.

File: python/fluid/gan.py (hunk 5 above)
Comment (wangkuiyi) [reply python/fluid/gan.py:42]: I am not sure but I feel that `a.func()` is more confusing than a series of function calls.
